bundle-workflow/src/test_workflow/test_recorder/test_recorder.py

```diff
@@ -0,0 +1,123 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# The OpenSearch Contributors require contributions made to
+# this file be licensed under the Apache-2.0 license or a
+# compatible open source license.
+
+import logging
+import os
+import shutil
+from typing import Dict
+
+import yaml
+
+from test_workflow.test_recorder.test_recorder_builder import \
+    TestRecorderBuilder
+
+
+class Singleton(type):
+    """Metaclass."""
+    _instances: Dict[type, object] = {}
+
+    def __call__(cls, *args, **kwargs):
+        if cls not in cls._instances:
+            cls._instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
+        else:
+            logging.info("TestRecorder is a singleton class, retrieving existing instance")
+        return cls._instances[cls]
+
+    def get_instance(cls):
+        if cls._instances is not None:
+            return cls._instances[cls]
+
+
+class TestRecorder(object, metaclass=Singleton):
+    ACCEPTABLE_TEST_TYPES = ["integ-test", "bwc-test", "perf-test"]
+
+    def __init__(self, test_run_id, test_type, location):
+        self.test_run_id = test_run_id
+        self.test_type = test_type
+        self.location = os.path.join(location, "test-recorder")
+        logging.info(f"TestRecorder recording logs in {self.location}")
+        if self.test_type not in self.ACCEPTABLE_TEST_TYPES:
+            raise ValueError(
+                f"{self.test_type} is not invalid. Accceptable test_type are {self.ACCEPTABLE_TEST_TYPES}")
+
+    def record_cluster_logs(self, cluster_type, test_recorder_builder: TestRecorderBuilder):
+        acceptable_cluster_type = ["remote", "local"]
+        if cluster_type not in acceptable_cluster_type:
+            raise ValueError(f"{cluster_type} is not invalid. Accceptable cluster_type are {acceptable_cluster_type}")
+        base = self.__create_base_folder_structure(test_recorder_builder.component_name,
+                                                   test_recorder_builder.component_test_config)
+        folder_name = cluster_type + "-cluster-logs"
+        dest_directory = os.path.join(base, folder_name)
+        os.makedirs(dest_directory, exist_ok=False)
+        logging.info(
+            f"Recording {cluster_type} cluster logs for {test_recorder_builder.component_name} "
+            f"with {test_recorder_builder.component_test_config} config in {os.path.realpath(dest_directory)}")
+        self.__generate_std_files(test_recorder_builder.stdout, test_recorder_builder.stderr,
+                                  os.path.realpath(dest_directory))
+        if cluster_type == "local":
+            local_cluster_log_files = list(test_recorder_builder.log_files)
+            for log_file in local_cluster_log_files:
```

Comment:
This becomes `LocalClusterLogs.write_to(...)` and `RemoteClusterLogs.write_to(...)`.